File: crates/kuben-store/src/repo/backups.rs
//! Backups, restores and the database's own facts (M4.7, migration 0025).
//!
//! After a restore the installation must not trust what happened after the
//! backup was taken: [`Store::after_restore`] ends every session, revokes
//! every API token, and raises every target's generation far beyond any the
//! clusters may have seen, so envelopes and objects written in the lost
//! interval are outranked instead of mistaken for newer state.

use std::time::Duration;

use kuben_core::{
    ids::{ConfigRevisionId, OrgId, ProjectId, ReleaseId, TargetId},
    ops::Generation,
    time::now_ms,
};
use sqlx::postgres::PgPoolOptions;
use uuid::Uuid;

use super::{NewAudit, RunReason, StartDeployment, Started, product::counter};
use crate::{Store, StoreError};

/// How far a restore raises every generation.
pub const RESTORE_GENERATION_JUMP: i64 = 1 << 20;

const INSERT_BACKUP: &str = "INSERT INTO backup_runs \
     (id, kind, status, location, bytes, sha256, kuben, schema, detail, started_at, finished_at) \
     VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)";
const LAST_GOOD: &str = "SELECT location, bytes, sha256, finished_at FROM backup_runs \
     WHERE status = 'succeeded' ORDER BY finished_at DESC LIMIT 1";
const SCHEMA: &str = "SELECT COALESCE(max(version), 0) FROM _sqlx_migrations WHERE success";
const TABLES: &str = "SELECT count(*) FROM information_schema.tables WHERE table_schema = current_schema()";
const FACTS: &str = "SELECT current_setting('server_version_num')::int, \
     COALESCE((SELECT ssl FROM pg_stat_ssl WHERE pid = pg_backend_pid()), false), \
     current_setting('wal_level'), current_setting('archive_mode'), pg_is_in_recovery()";
const END_SESSIONS: &str = "UPDATE sessions SET revoked_at = $1 WHERE revoked_at IS NULL";
const REVOKE_TOKENS: &str = "UPDATE api_tokens SET revoked_at = $1 WHERE revoked_at IS NULL";
const RAISE_GENERATIONS: &str = "UPDATE application_targets \
     SET desired_generation = desired_generation + $2 WHERE org_id = $1";
/// Every live target with the release and configuration of its newest run.
const REDELIVERABLE: &str = "SELECT t.project_id, t.id, r.release_id, r.config_revision_id, \
     t.desired_generation, t.lifecycle_uid FROM application_targets t \
     JOIN LATERAL (SELECT d.release_id, d.config_revision_id FROM deployment_runs d \
       WHERE d.target_id = t.id AND d.org_id = t.org_id ORDER BY d.generation DESC LIMIT 1) r ON TRUE \
     WHERE t.org_id = $1 AND NOT t.deleting";
const INSERT_RESTORE: &str = "INSERT INTO restore_events \
     (id, backup_created_at, backup_kuben, kuben, generation_jump, restored_at) \
     VALUES ($1, $2, $3, $4, $5, $6)";

/// A backup written.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BackupRecord {
    pub scheduled: bool,
    pub succeeded: bool,
    pub location: String,
    pub bytes: Option<u64>,
    pub sha256: Option<String>,
    pub kuben: String,
    pub schema: i64,
    pub detail: Option<String>,
    pub started_at: i64,
    pub finished_at: i64,
}

/// The newest good backup.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct LastBackup {
    pub location: String,
    pub bytes: Option<u64>,
    pub sha256: Option<String>,
    pub finished_at: i64,
}

/// What the database server says about itself.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct DatabaseFacts {
    /// `server_version_num`, e.g. 170004.
    pub version_num: i32,
    /// This connection is encrypted.
    pub tls: bool,
    pub wal_level: String,
    pub archive_mode: String,
    /// A standby: Kuben cannot write here.
    pub in_recovery: bool,
}

impl DatabaseFacts {
    /// The major version, e.g. 17.
    #[must_use]
    pub const fn major(&self) -> i32 {
        self.version_num / 10_000
    }

    /// WAL is archived, so a point-in-time recovery is possible.
    #[must_use]
    pub fn archives_wal(&self) -> bool {
        matches!(self.archive_mode.as_str(), "on" | "always") && self.wal_level != "minimal"
    }
}

/// What a restore changed.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Restored {
    pub sessions_ended: u64,
    pub tokens_revoked: u64,
    pub targets_raised: u64,
}

impl Store {
    /// Whether the database at `url` has no tables yet (never migrated).
    pub async fn database_is_empty(url: &str) -> Result<bool, StoreError> {
        let pool = PgPoolOptions::new()
            .max_connections(1)
            .acquire_timeout(Duration::from_secs(10))
            .connect(url)
            .await?;
        let tables: i64 = sqlx::query_scalar(TABLES).fetch_one(&pool).await?;
        pool.close().await;
        Ok(tables == 0)
    }

    /// The newest migration applied.
    pub async fn schema_version(&self) -> Result<i64, StoreError> {
        Ok(sqlx::query_scalar(SCHEMA).fetch_one(self.pool()).await?)
    }

    /// The database server's version, encryption and WAL archiving.
    pub async fn database_facts(&self) -> Result<DatabaseFacts, StoreError> {
        let (version_num, tls, wal_level, archive_mode, in_recovery): (i32, bool, String, String, bool) =
            sqlx::query_as(FACTS).fetch_one(self.pool()).await?;
        Ok(DatabaseFacts {
            version_num,
            tls,
            wal_level,
            archive_mode,
            in_recovery,
        })
    }

    /// Record a backup attempt.
    pub async fn record_backup(&self, b: &BackupRecord) -> Result<(), StoreError> {
        let bytes = b
            .bytes
            .map(i64::try_from)
            .transpose()
            .map_err(|e| sqlx::Error::Encode(e.into()))?;
        sqlx::query(INSERT_BACKUP)
            .bind(Uuid::now_v7())
            .bind(if b.scheduled { "scheduled" } else { "manual" })
            .bind(if b.succeeded { "succeeded" } else { "failed" })
            .bind(&b.location)
            .bind(bytes)
            .bind(b.sha256.as_deref())
            .bind(&b.kuben)
            .bind(b.schema)
            .bind(b.detail.as_deref())
            .bind(b.started_at)
            .bind(b.finished_at)
            .execute(self.pool())
            .await?;
        Ok(())
    }

    /// The newest good backup, if any.
    pub async fn last_backup(&self) -> Result<Option<LastBackup>, StoreError> {
        let row: Option<(String, Option<i64>, Option<String>, i64)> =
            sqlx::query_as(LAST_GOOD).fetch_optional(self.pool()).await?;
        row.map(|(location, bytes, sha256, finished_at)| {
            Ok(LastBackup {
                location,
                bytes: bytes.map(counter).transpose()?,
                sha256,
                finished_at,
            })
        })
        .transpose()
    }

    /// Fence everything that happened after the restored backup was taken,
    /// in one transaction per step: sessions, API tokens, and generations of
    /// every organization's targets.
    pub async fn after_restore(
        &self,
        backup_created_at: i64,
        backup_kuben: &str,
        kuben: &str,
    ) -> Result<Restored, StoreError> {
        let now = now_ms();
        let mut tx = self.pool().begin().await?;
        let sessions_ended = sqlx::query(END_SESSIONS)
            .bind(now)
            .execute(&mut *tx)
            .await?
            .rows_affected();
        let tokens_revoked = sqlx::query(REVOKE_TOKENS)
            .bind(now)
            .execute(&mut *tx)
            .await?
            .rows_affected();
        sqlx::query(INSERT_RESTORE)
            .bind(Uuid::now_v7())
            .bind(backup_created_at)
            .bind(backup_kuben)
            .bind(kuben)
            .bind(RESTORE_GENERATION_JUMP)
            .bind(now)
            .execute(&mut *tx)
            .await?;
        tx.commit().await?;
        let mut targets_raised = 0;
        for org in self.org_ids().await? {
            targets_raised += self.raise_generations(org).await?;
        }
        Ok(Restored {
            sessions_ended,
            tokens_revoked,
            targets_raised,
        })
    }

    /// Start a run of every live target's newest release and configuration,
    /// so the restored state is written again (to a fresh cluster, too).
    /// Runs are restarts: they carry nothing new, so neither approvals nor
    /// the scan gate hold them. The number started.
    pub async fn redeliver_all(&self, requested_by: &str) -> Result<usize, StoreError> {
        let mut started = 0;
        for org in self.org_ids().await? {
            let mut tenant = self.tenant(org).await?;
            let rows: Vec<(Uuid, Uuid, Uuid, Uuid, i64, Uuid)> = sqlx::query_as(REDELIVERABLE)
                .bind(org.to_string())
                .fetch_all(&mut *tenant.tx)
                .await?;
            for (project, target, release, config, generation, lifecycle_uid) in rows {
                let req = StartDeployment {
                    project: ProjectId::from_uuid(project),
                    target: TargetId::from_uuid(target),
                    release: ReleaseId::from_uuid(release),
                    config_revision: ConfigRevisionId::from_uuid(config),
                    render_plan: None,
                    expected_generation: Generation(counter(generation)?),
                    lifecycle_uid,
                    reason: RunReason::Restart,
                    requested_by: requested_by.to_owned(),
                    input_hash: format!("restore/{target}/{generation}").into_bytes(),
                };
                let audit = NewAudit {
                    actor_kind: "system".into(),
                    actor_id: Some(requested_by.to_owned()),
                    action: "deployment.redelivered".into(),
                    target_kind: Some("app".into()),
                    target_ref: Some(target.to_string()),
                    outcome: "accepted".into(),
                    ..NewAudit::default()
                };
                if matches!(
                    tenant.start_deployment(&req, audit, None).await?,
                    Started::Accepted { .. }
                ) {
                    started += 1;
                }
            }
            tenant.commit().await?;
        }
        Ok(started)
    }

    async fn raise_generations(&self, org: OrgId) -> Result<u64, StoreError> {
        let mut tenant = self.tenant(org).await?;
        let raised = sqlx::query(RAISE_GENERATIONS)
            .bind(org.to_string())
            .bind(RESTORE_GENERATION_JUMP)
            .execute(&mut *tenant.tx)
            .await?
            .rows_affected();
        tenant.commit().await?;
        Ok(raised)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::{pg_store, skip};

    #[test]
    fn wal_archiving_needs_a_mode_and_a_level() {
        let facts = |level: &str, mode: &str| DatabaseFacts {
            version_num: 170_004,
            tls: false,
            wal_level: level.into(),
            archive_mode: mode.into(),
            in_recovery: false,
        };
        assert_eq!(facts("replica", "on").major(), 17);
        assert!(facts("replica", "on").archives_wal());
        assert!(facts("logical", "always").archives_wal());
        assert!(!facts("replica", "off").archives_wal());
        assert!(!facts("minimal", "on").archives_wal());
    }

    #[tokio::test]
    async fn backups_are_recorded_and_restores_fence_the_past() {
        let Some(store) = pg_store().await else {
            return skip("backups_are_recorded_and_restores_fence_the_past");
        };
        assert!(store.schema_version().await.expect("schema") >= 25);
        let facts = store.database_facts().await.expect("facts");
        assert!(facts.major() >= 15 && !facts.in_recovery);
        assert_eq!(store.last_backup().await.expect("read"), None);
        let record = |succeeded: bool, at: i64| BackupRecord {
            scheduled: true,
            succeeded,
            location: format!("/backups/{at}"),
            bytes: succeeded.then_some(1024),
            sha256: succeeded.then(|| "a".repeat(64)),
            kuben: "2.0.0".into(),
            schema: 25,
            detail: (!succeeded).then(|| "pg_dump failed".into()),
            started_at: at - 5,
            finished_at: at,
        };
        store.record_backup(&record(true, 100)).await.expect("backup");
        store.record_backup(&record(false, 200)).await.expect("backup");
        let last = store.last_backup().await.expect("read").expect("backup");
        assert_eq!(
            (last.location.as_str(), last.bytes, last.finished_at),
            ("/backups/100", Some(1024), 100)
        );

        let org = store.create_org("a", "A").await.expect("org").id;
        let mut t = store.tenant(org).await.expect("tenant");
        let project = t.create_project("shop", "Shop").await.expect("project");
        let env = t
            .create_environment(project, "prod", "Prod", false)
            .await
            .expect("environment");
        let cluster = t.create_cluster("eu-1").await.expect("cluster");
        let placement = t
            .create_placement(project, env, cluster, "a-shop")
            .await
            .expect("placement");
        let application = t.create_application(project, "web", "Web").await.expect("app");
        let target = t
            .create_target(project, application, placement)
            .await
            .expect("target");
        t.commit().await.expect("commit");

        let restored = store.after_restore(50, "1.9.0", "2.0.0").await.expect("restore");
        assert_eq!(restored.targets_raised, 1);
        let mut t = store.tenant(org).await.expect("tenant");
        let state = t.target_state(target).await.expect("read").expect("target");
        assert_eq!(
            state.desired_generation.0,
            u64::try_from(RESTORE_GENERATION_JUMP).expect("jump")
        );
    }
}

File: go/hub/internal/store/backups.go
package store

// Backups, restores and the database's own facts (M4.7, migration 0025);
// the port of repo/backups.rs.
//
// After a restore the installation must not trust what happened after the
// backup was taken: [Store.AfterRestore] ends every session, revokes every
// API token, and raises every target's generation far beyond any the
// clusters may have seen, so envelopes and objects written in the lost
// interval are outranked instead of mistaken for newer state.

import (
	"context"
	"errors"
	"fmt"
	"math"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"

	"github.com/Teamtem-dev/kuben/go/hub/internal/core/ids"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/ops/target"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/opt"
)

// RestoreGenerationJump is how far a restore raises every generation.
const RestoreGenerationJump int64 = 1 << 20

const (
	insertBackup = "INSERT INTO backup_runs " +
		"(id, kind, status, location, bytes, sha256, kuben, schema, detail, started_at, finished_at) " +
		"VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)"
	lastGoodBackup = "SELECT location, bytes, sha256, finished_at FROM backup_runs " +
		"WHERE status = 'succeeded' ORDER BY finished_at DESC LIMIT 1"
	schemaVersion = "SELECT COALESCE(max(version), 0) FROM _sqlx_migrations WHERE success"
	countTables   = "SELECT count(*) FROM information_schema.tables WHERE table_schema = current_schema()"
	databaseFacts = "SELECT current_setting('server_version_num')::int, " +
		"COALESCE((SELECT ssl FROM pg_stat_ssl WHERE pid = pg_backend_pid()), false), " +
		"current_setting('wal_level'), current_setting('archive_mode'), pg_is_in_recovery()"
	endSessions      = "UPDATE sessions SET revoked_at = $1 WHERE revoked_at IS NULL"
	revokeAllTokens  = "UPDATE api_tokens SET revoked_at = $1 WHERE revoked_at IS NULL"
	raiseGenerations = "UPDATE application_targets " +
		"SET desired_generation = desired_generation + $2 WHERE org_id = $1"
	// redeliverable is every live target with the release and
	// configuration of its newest run.
	redeliverable = "SELECT t.project_id, t.id, r.release_id, r.config_revision_id, " +
		"t.desired_generation, t.lifecycle_uid FROM application_targets t " +
		"JOIN LATERAL (SELECT d.release_id, d.config_revision_id FROM deployment_runs d " +
		"WHERE d.target_id = t.id AND d.org_id = t.org_id ORDER BY d.generation DESC LIMIT 1) r ON TRUE " +
		"WHERE t.org_id = $1 AND NOT t.deleting"
	insertRestore = "INSERT INTO restore_events " +
		"(id, backup_created_at, backup_kuben, kuben, generation_jump, restored_at) " +
		"VALUES ($1, $2, $3, $4, $5, $6)"
)

// BackupRecord is a backup written.
type BackupRecord struct {
	Scheduled  bool
	Succeeded  bool
	Location   string
	Bytes      opt.Val[uint64]
	SHA256     opt.Val[string]
	Kuben      string
	Schema     int64
	Detail     opt.Val[string]
	StartedAt  int64
	FinishedAt int64
}

// LastBackup is the newest good backup.
type LastBackup struct {
	Location   string
	Bytes      opt.Val[uint64]
	SHA256     opt.Val[string]
	FinishedAt int64
}

// DatabaseFacts is what the database server says about itself.
type DatabaseFacts struct {
	// VersionNum is `server_version_num`, e.g. 170004.
	VersionNum int32
	// TLS: this connection is encrypted.
	TLS         bool
	WALLevel    string
	ArchiveMode string
	// InRecovery: a standby, where Kuben cannot write.
	InRecovery bool
}

// Major is the major version, e.g. 17.
func (f DatabaseFacts) Major() int32 { return f.VersionNum / 10_000 }

// ArchivesWAL reports whether WAL is archived, so a point-in-time recovery
// is possible.
func (f DatabaseFacts) ArchivesWAL() bool {
	return (f.ArchiveMode == "on" || f.ArchiveMode == "always") && f.WALLevel != "minimal"
}

// Restored is what a restore changed.
type Restored struct {
	SessionsEnded uint64
	TokensRevoked uint64
	TargetsRaised uint64
}

// DatabaseIsEmpty reports whether the database at url has no tables yet
// (never migrated).
func DatabaseIsEmpty(ctx context.Context, url string) (bool, error) {
	const op = "check the database is empty"
	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()
	conn, err := pgx.Connect(ctx, url)
	if err != nil {
		return false, dbErr(op, err)
	}
	var tables int64
	err = conn.QueryRow(ctx, countTables).Scan(&tables)
	return tables == 0, errors.Join(dbErr(op, err), dbErr(op, conn.Close(ctx)))
}

// SchemaVersion is the newest migration applied.
func (s *Store) SchemaVersion(ctx context.Context) (int64, error) {
	var v int64
	return v, dbErr("read the schema version", s.db.QueryRow(ctx, schemaVersion).Scan(&v))
}

// DatabaseFacts is the database server's version, encryption and WAL
// archiving.
func (s *Store) DatabaseFacts(ctx context.Context) (DatabaseFacts, error) {
	var f DatabaseFacts
	err := s.db.QueryRow(ctx, databaseFacts).Scan(&f.VersionNum, &f.TLS, &f.WALLevel, &f.ArchiveMode, &f.InRecovery)
	return f, dbErr("read the database facts", err)
}

// RecordBackup records a backup attempt.
func (s *Store) RecordBackup(ctx context.Context, b BackupRecord) error {
	const op = "record a backup"
	bytes := opt.None[int64]()
	if n, ok := b.Bytes.Get(); ok {
		if n > math.MaxInt64 {
			return protocolErr(op, "a backup of %d bytes does not fit a bigint", n)
		}
		bytes = opt.Some(int64(n))
	}
	kind, status := "manual", "failed"
	if b.Scheduled {
		kind = "scheduled"
	}
	if b.Succeeded {
		status = "succeeded"
	}
	id, err := uuid.NewV7()
	if err != nil {
		return fmt.Errorf("%s: %w", op, err)
	}
	_, err = exec(ctx, s.db, op, insertBackup, id, kind, status, b.Location, bytes.Ptr(), b.SHA256.Ptr(),
		b.Kuben, b.Schema, b.Detail.Ptr(), b.StartedAt, b.FinishedAt)
	return err
}

// LastBackup is the newest good backup, if any.
func (s *Store) LastBackup(ctx context.Context) (LastBackup, bool, error) {
	const op = "read the last backup"
	var (
		last   LastBackup
		bytes  *int64
		sha256 *string
	)
	err := s.db.QueryRow(ctx, lastGoodBackup).Scan(&last.Location, &bytes, &sha256, &last.FinishedAt)
	if errors.Is(err, pgx.ErrNoRows) {
		return LastBackup{}, false, nil
	}
	if err != nil {
		return LastBackup{}, false, dbErr(op, err)
	}
	if bytes != nil {
		n, err := counter(op, *bytes)
		if err != nil {
			return LastBackup{}, false, err
		}
		last.Bytes = opt.Some(n)
	}
	last.SHA256 = opt.FromPtr(sha256)
	return last, true, nil
}

// AfterRestore fences everything that happened after the restored backup
// was taken, in one transaction per step: sessions, API tokens, and the
// generations of every organization's targets.
func (s *Store) AfterRestore(ctx context.Context, backupCreatedAt int64, backupKuben, kuben string) (Restored, error) {
	const op = "fence a restore"
	now := s.now()
	var r Restored
	tx, err := s.db.Begin(ctx)
	if err != nil {
		return r, dbErr(op, err)
	}
	defer func() { _ = rollback(ctx, tx) }() //nolint:errcheck // committed below on success
	if r.SessionsEnded, err = exec(ctx, tx, op, endSessions, now); err != nil {
		return r, err
	}
	if r.TokensRevoked, err = exec(ctx, tx, op, revokeAllTokens, now); err != nil {
		return r, err
	}
	id, err := uuid.NewV7()
	if err != nil {
		return r, fmt.Errorf("%s: %w", op, err)
	}
	if _, err := exec(ctx, tx, op, insertRestore, id, backupCreatedAt, backupKuben, kuben, RestoreGenerationJump, now); err != nil {
		return r, err
	}
	if err := tx.Commit(ctx); err != nil {
		return r, dbErr(op, err)
	}
	orgs, err := s.OrgIDs(ctx)
	if err != nil {
		return r, err
	}
	for _, org := range orgs {
		raised, err := s.raiseGenerations(ctx, org)
		if err != nil {
			return r, err
		}
		r.TargetsRaised += raised
	}
	return r, nil
}

func (s *Store) raiseGenerations(ctx context.Context, org ids.OrgID) (uint64, error) {
	t, err := s.Tenant(ctx, org)
	if err != nil {
		return 0, err
	}
	defer t.Rollback(ctx) //nolint:errcheck // committed below on success
	raised, err := exec(ctx, t.tx, "raise the generations", raiseGenerations, org.String(), RestoreGenerationJump)
	if err != nil {
		return 0, err
	}
	return raised, t.Commit(ctx)
}

// RedeliverAll starts a run of every live target's newest release and
// configuration, so the restored state is written again (to a fresh
// cluster, too). Runs are restarts: they carry nothing new, so neither
// approvals nor the scan gate hold them. It is the number started.
func (s *Store) RedeliverAll(ctx context.Context, requestedBy string) (int, error) {
	orgs, err := s.OrgIDs(ctx)
	if err != nil {
		return 0, err
	}
	started := 0
	for _, org := range orgs {
		n, err := s.redeliver(ctx, org, requestedBy)
		if err != nil {
			return started, err
		}
		started += n
	}
	return started, nil
}

// redeliverRow is a row of the redeliverable query.
type redeliverRow struct {
	project, target, release, config, lifecycle uuid.UUID
	generation                                  int64
}

func (s *Store) redeliver(ctx context.Context, org ids.OrgID, requestedBy string) (int, error) {
	const op = "redeliver the targets"
	t, err := s.Tenant(ctx, org)
	if err != nil {
		return 0, err
	}
	defer t.Rollback(ctx) //nolint:errcheck // committed below on success
	rows, err := queryAll(ctx, t.tx, op, redeliverable, func(r pgx.CollectableRow) (redeliverRow, error) {
		var x redeliverRow
		return x, r.Scan(&x.project, &x.target, &x.release, &x.config, &x.generation, &x.lifecycle)
	}, org.String())
	if err != nil {
		return 0, err
	}
	started := 0
	for _, row := range rows {
		generation, err := counter(op, row.generation)
		if err != nil {
			return 0, err
		}
		req := StartDeployment{
			Project: ids.From[ids.Project](row.project), Target: ids.From[ids.Target](row.target),
			Release: ids.From[ids.Release](row.release), ConfigRevision: ids.From[ids.ConfigRevision](row.config),
			ExpectedGeneration: target.Generation(generation), LifecycleUID: row.lifecycle,
			Reason: ReasonRestart, RequestedBy: requestedBy,
			InputHash: fmt.Appendf(nil, "restore/%s/%d", row.target, row.generation),
		}
		audit := NewAudit{
			ActorKind: "system", ActorID: opt.Some(requestedBy), Action: "deployment.redelivered",
			TargetKind: opt.Some("app"), TargetRef: opt.Some(row.target.String()), Outcome: "accepted",
		}
		result, err := t.StartDeployment(ctx, req, audit, opt.None[IdempotencyKey]())
		if err != nil {
			return 0, err
		}
		if _, ok := result.(StartedAccepted); ok {
			started++
		}
	}
	return started, t.Commit(ctx)
}
